File: identitykit/src/test/java/com/eldersoss/identitykit/ClientCredentialsFlowTest.java
package com.eldersoss.identitykit;

import com.eldersoss.identitykit.authorization.Authorizer;
import com.eldersoss.identitykit.authorization.BasicAuthorizer;
import com.eldersoss.identitykit.authorization.BearerAuthorizer;
import com.eldersoss.identitykit.network.NetworkClient;
import com.eldersoss.identitykit.network.NetworkRequest;
import com.eldersoss.identitykit.network.NetworkResponse;
import com.eldersoss.identitykit.oauth2.DefaultTokenRefresher;
import com.eldersoss.identitykit.oauth2.Error;
import com.eldersoss.identitykit.oauth2.OAuth2Error;
import com.eldersoss.identitykit.oauth2.flows.AuthorizationFlow;
import com.eldersoss.identitykit.oauth2.flows.ClientCredentialsFlow;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowLooper;

import java.util.HashMap;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;

import static com.eldersoss.identitykit.storage.ConstantsKt.REFRESH_TOKEN;
import static org.junit.Assert.assertTrue;

/**
 * Created by IvanVatov on 9/4/2017.
 */
@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class)
public class ClientCredentialsFlowTest {

    @Test
    public void identityKitInitializationTest() throws Exception {
        IdentityKit kit;
        NetworkClient networkClient;

        networkClient = new TestNetworkClient();
        Authorizer authorizer = new BasicAuthorizer("client", "secret");
        AuthorizationFlow flow = new ClientCredentialsFlow("https://account.foo.bar/token", "read write openid email profile offline_access owner", authorizer, networkClient);
        kit = new IdentityKit(flow, BearerAuthorizer.Method.HEADER, new DefaultTokenRefresher("https://account.foo.bar/token", networkClient, authorizer), new TestTokenStorage(), networkClient);

        kit.authorizeAndExecute(new NetworkRequest("GET", "https://account.foo.bar/profile", new HashMap<String, String>(), "".getBytes()), new Function1<NetworkResponse, Unit>() {
            @Override
            public Unit invoke(NetworkResponse networkResponse) {
                return null;
            }
        });
        assertTrue(kit != null);
    }

    @Test
    public void successAuthorizeTest() throws Exception {
        IdentityKit kit;
        NetworkClient networkClient;
        networkClient = new TestNetworkClient();

        // test stuffs
        final TestResultHandler handler = new TestResultHandler();
        ((TestNetworkClient) networkClient).setCase(TestNetworkClient.ResponseCase.OK200);

        Authorizer authorizer = new BasicAuthorizer("client", "secret");
        AuthorizationFlow flow = new ClientCredentialsFlow("https://account.foo.bar/token", "read write openid email profile offline_access owner", authorizer, networkClient);
        kit = new IdentityKit(flow, BearerAuthorizer.Method.HEADER, new DefaultTokenRefresher("https://account.foo.bar/token", networkClient, authorizer), new TestTokenStorage(), networkClient);

        final NetworkRequest request = new NetworkRequest("GET", "https://account.foo.bar/api/profile", new HashMap<String, String>(), "".getBytes());
        kit.authorize(request, new Function2<NetworkRequest, Error, Unit>() {
            @Override
            public Unit invoke(NetworkRequest networkRequest, Error error) {
                handler.value = networkRequest.getHeaders().get("Authorization");
                return null;
            }
        });
        ShadowLooper.runUiThreadTasksIncludingDelayedTasks();
        String responseAuthorization = "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9";
        assertTrue(handler.value.equalsIgnoreCase(responseAuthorization));
    }

    @Test
    public void invalidGrandTest() throws Exception {
        IdentityKit kit;
        NetworkClient networkClient;
        networkClient = new TestNetworkClient();

        // test stuffs
        final TestResultHandler handler = new TestResultHandler();
        ((TestNetworkClient) networkClient).setCase(TestNetworkClient.ResponseCase.BAD400);

        // building IdentityKit
        Authorizer authorizer = new BasicAuthorizer("client", "secret");
        AuthorizationFlow flow = new ClientCredentialsFlow("https://account.foo.bar/token", "read write openid email profile offline_access owner", authorizer, networkClient);
        kit = new IdentityKit(flow, BearerAuthorizer.Method.HEADER, new DefaultTokenRefresher("https://account.foo.bar/token", networkClient, authorizer), new TestTokenStorage(), networkClient);

        final NetworkRequest request = new NetworkRequest("GET", "https://account.foo.bar/api/profile", new HashMap<String, String>(), "".getBytes());
        kit.authorize(request, new Function2<NetworkRequest, Error, Unit>() {
            @Override
            public Unit invoke(NetworkRequest networkRequest, Error error) {
                handler.error = error;
                return null;
            }
        });
        ShadowLooper.runUiThreadTasksIncludingDelayedTasks();
        assertTrue(handler.error.equals(OAuth2Error.invalid_grant));
    }

    @Test
    public void refreshToken() throws Exception {
        IdentityKit kit;
        NetworkClient networkClient;
        TestTokenStorage tokenStorage = new TestTokenStorage();
        networkClient = new TestNetworkClient();

        // test stuffs
        final TestResultHandler handler = new TestResultHandler();
        tokenStorage.write(REFRESH_TOKEN, "4f2aw4gf5ge0c3aa3as2e4f8a958c6");
        ((TestNetworkClient) networkClient).setCase(TestNetworkClient.ResponseCase.REFRESH200);

        Authorizer authorizer = new BasicAuthorizer("client", "secret");
        AuthorizationFlow flow = new ClientCredentialsFlow("https://account.foo.bar/token", "read write openid email profile offline_access owner", authorizer, networkClient);
        kit = new IdentityKit(flow, BearerAuthorizer.Method.HEADER, new DefaultTokenRefresher("https://account.foo.bar/token", networkClient, authorizer), tokenStorage, networkClient);

        final NetworkRequest request = new NetworkRequest("GET", "https://account.foo.bar/api/profile", new HashMap<String, String>(), "".getBytes());
        kit.authorize(request, new Function2<NetworkRequest, Error, Unit>() {
            @Override
            public Unit invoke(NetworkRequest networkRequest, Error error) {
                handler.value = networkRequest.getHeaders().get("Authorization");
                return null;
            }
        });
        ShadowLooper.runUiThreadTasksIncludingDelayedTasks();

        String responseAuthorization = "Bearer TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ";
        assertTrue(handler.value.equalsIgnoreCase(responseAuthorization));
    }
}

File: identitykit/src/test/java/com/eldersoss/identitykit/ClientCredentialsFlowTest.kt
package com.eldersoss.identitykit

import com.eldersoss.identitykit.authorization.BasicAuthorizer
import com.eldersoss.identitykit.authorization.BearerAuthorizer
import com.eldersoss.identitykit.exceptions.OAuth2Exception
import com.eldersoss.identitykit.exceptions.OAuth2InvalidGrand
import com.eldersoss.identitykit.network.NetworkClient
import com.eldersoss.identitykit.network.NetworkRequest
import com.eldersoss.identitykit.oauth2.DefaultTokenRefresher
import com.eldersoss.identitykit.oauth2.flows.ClientCredentialsFlow
import com.eldersoss.identitykit.storage.REFRESH_TOKEN
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

/**
 * Created by IvanVatov on 11/7/2017.
 */
@ExperimentalCoroutinesApi
@RunWith(RobolectricTestRunner::class)
class ClientCredentialsFlowTest {

    private val configuration = KitConfiguration(
        retryFlowAuthentication = false,
        authenticateOnFailedRefresh = false,
        onAuthenticationRetryInvokeCallbackWithFailure = false
    )

    @Test
    fun identityKitInitializationTest() = runBlockingTest {
        val kit: IdentityKit?
        val networkClient: NetworkClient

        networkClient = MockNetworkClient()
        networkClient.setCase(MockNetworkClient.ResponseCase.CC200OK)

        val authorizer = BasicAuthorizer("client", "secret")
        val flow = ClientCredentialsFlow(
            "https://account.foo.bar/token",
            "read write openid email profile offline_access owner",
            authorizer,
            networkClient
        )

        kit = IdentityKit(
            configuration,
            flow,
            BearerAuthorizer.Method.HEADER,
            DefaultTokenRefresher("https://account.foo.bar/token", networkClient, authorizer),
            TestTokenStorage(),
            networkClient,
            Dispatchers.Unconfined
        )

        kit.authorizeAndExecute(
            NetworkRequest(
                NetworkRequest.Method.GET,
                NetworkRequest.Priority.HIGH,
                "https://account.foo.bar/profile"
            )
        )

        Assert.assertTrue(kit != null)
    }

    /**
     * Unit test on asynchronous implementation in this depth cannot be provided by currently available test instruments
     */
    @Test
    fun successAuthorizeTest() = runBlockingTest {

        val kit: IdentityKit
        val networkClient: NetworkClient
        networkClient = MockNetworkClient()
        networkClient.setCase(MockNetworkClient.ResponseCase.CC200OK)

        val authorizer = BasicAuthorizer("client", "secret")
        val flow = ClientCredentialsFlow(
            "https://account.foo.bar/token",
            "read write openid email profile offline_access owner",
            authorizer,
            networkClient
        )
        kit = IdentityKit(
            configuration,
            flow,
            BearerAuthorizer.Method.HEADER,
            DefaultTokenRefresher("https://account.foo.bar/token", networkClient, authorizer),
            TestTokenStorage(),
            networkClient,
            Dispatchers.Unconfined
        )

        val request = NetworkRequest(
            NetworkRequest.Method.GET,
            NetworkRequest.Priority.HIGH,
            "https://account.foo.bar/api/profile"
        )

        kit.authorize(request)

        val authHeaderValue = request.headers["Authorization"]
        val responseAuthorization = "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9"

        Assert.assertTrue(authHeaderValue.equals(responseAuthorization))
    }

    @Test
    fun invalidGrandTest() = runBlockingTest {

        val kit: IdentityKit
        val networkClient: NetworkClient
        networkClient = MockNetworkClient()
        networkClient.setCase(MockNetworkClient.ResponseCase.BAD400)

        // building IdentityKit
        val authorizer = BasicAuthorizer("client", "secret")
        val flow = ClientCredentialsFlow(
            "https://account.foo.bar/token",
            "read write openid email profile offline_access owner",
            authorizer,
            networkClient
        )
        kit = IdentityKit(
            configuration,
            flow,
            BearerAuthorizer.Method.HEADER,
            DefaultTokenRefresher("https://account.foo.bar/token", networkClient, authorizer),
            TestTokenStorage(),
            networkClient,
            Dispatchers.Unconfined
        )

        val request = NetworkRequest(
            NetworkRequest.Method.GET,
            NetworkRequest.Priority.HIGH,
            "https://account.foo.bar/api/profile"
        )

        var oauth2Exception: OAuth2Exception? = null

        try {

            kit.authorize(request)
        } catch (e: OAuth2InvalidGrand) {

            oauth2Exception = e
        }

        Assert.assertTrue(oauth2Exception is OAuth2InvalidGrand)
    }

    /**
     * Unit test on asynchronous implementation in this depth cannot be provided by currently available test instruments
     */
    @Test
    fun refreshToken() = runBlockingTest {

        val kit: IdentityKit
        val networkClient: NetworkClient
        val tokenStorage = TestTokenStorage()
        networkClient = MockNetworkClient()
        tokenStorage.write(REFRESH_TOKEN, "4f2aw4gf5ge0c3aa3as2e4f8a958c6")
        networkClient.setCase(MockNetworkClient.ResponseCase.REFRESH200)

        val authorizer = BasicAuthorizer("client", "secret")
        val flow = ClientCredentialsFlow(
            "https://account.foo.bar/token",
            "read write openid email profile offline_access owner",
            authorizer,
            networkClient
        )
        kit = IdentityKit(
            configuration,
            flow,
            BearerAuthorizer.Method.HEADER,
            DefaultTokenRefresher("https://account.foo.bar/token", networkClient, authorizer),
            tokenStorage,
            networkClient,
            Dispatchers.Unconfined
        )

        val request = NetworkRequest(
            NetworkRequest.Method.GET,
            NetworkRequest.Priority.HIGH,
            "https://account.foo.bar/api/profile"
        )

        kit.authorize(request)

        val authHeaderValue = request.headers["Authorization"]
        val responseAuthorization = "Bearer TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ"

        Assert.assertTrue(authHeaderValue.equals(responseAuthorization))
    }

}